File: src/lib/tiposkripto/src/BaseValidate.ts
import { BaseCheck } from "./BaseCheck.js";
import { TestTypeParams_any } from "./CoreTypes.js";
import { ITestResourceConfiguration } from "./types.js";

/**
 * BaseValidate extends BaseCheck to support TDT (Table Driven Testing) pattern.
 * It validates the output against expected results.
 */
export class BaseValidate<I extends TestTypeParams_any> extends BaseCheck<I> {
  // Expected result for the current row
  expectedResult: any = null;

  constructor(
    name: string,
    validateCB: (val: I["iselection"]) => Promise<I["then"]>
  ) {
    super(name, validateCB);
  }

  // Set expected result before validation
  setExpectedResult(expected: any) {
    this.expectedResult = expected;
  }

  // Alias verifyCheck to validate for TDT pattern
  async validate(
    store: I["istore"],
    validateCB: (s: I["iselection"]) => Promise<I["isubject"]>,
    testResourceConfiguration: ITestResourceConfiguration
  ) {
    return super.verifyCheck(store, validateCB, testResourceConfiguration);
  }

  // Alias test to check for TDT pattern
  async check(
    store: I["istore"],
    testResourceConfiguration,
    filepath: string,
    expectedResult: any
  ) {
    this.setExpectedResult(expectedResult);
    return super.test(store, testResourceConfiguration, filepath);
  }
}

export type IValidates<I extends TestTypeParams_any> = Record<string, BaseValidate<I>>;

File: src/lib/rubeno/lib/base_validate.rb
module Rubeno
  # BaseValidate extends BaseCheck to support TDT (Table Driven Testing) pattern.
  class BaseValidate < BaseCheck
    attr_accessor :expected_result
    
    def initialize(name, validate_cb)
      super(name, validate_cb)
      @expected_result = nil
    end
    
    # Set expected result before validation
    def set_expected_result(expected)
      @expected_result = expected
    end
    
    # Alias verify_check to validate for TDT pattern
    def validate(store, validate_cb, test_resource_configuration)
      verify_check(store, validate_cb, test_resource_configuration)
    end
    
    # Alias test to check for TDT pattern
    def check(store, test_resource_configuration, filepath, expected_result)
      set_expected_result(expected_result)
      test(store, test_resource_configuration, filepath)
    end
  end
end
